visions/abi-stability.md

```diff
@@ -0,0 +1,248 @@
+# A Vision for ABI Stability on non-Apple platforms
+
+## Introduction
+
+ABI or _Application Binary Interface_ refers to the details of how an
+application will interact with some other piece of code, including:
+
+* Calling conventions
+* Layout of data structures
+* Representations of scalar types
+* Availability of APIs
+
+If a piece of code has a _stable ABI_, that means that application
+programs can rely on there being no incompatible changes between the
+state of the ABI when they were being compiled and the ABI provided at
+runtime by the component they were linked with.
+
+End users of consumer operating systems, including both Apple and
+Microsoft platforms, tend to take ABI stability for granted; that is,
```

Comment:
Is it as true for Microsoft/Windows? I know they're moving toward a "universal C runtime" so this may not be as true in the future (though certainly still true for the C++ runtime), but you still end up bundling the language runtime DLLs in the application bundle next to the executable so it's less ABI stability and more, "you pack the runtime you built against so there aren't ABI conflicts"? Sort of like snaps or how we used to ship the Swift standard library on Apple platforms before we had ABI stability.

Reply:
:-) It's complicated.

The C runtime on Windows isn't tightly coupled with the OS in the way that the C library on Darwin or Linux is; system components don't use the same C runtime library that applications use, and it's totally possible to write programs that just use the Win32 API and don't use the C runtime at all (or very little, at any rate).

So yes, programs will often install a C runtime library somewhere (hopefully, these days, not in `C:\Windows\System32`) and not share that with other programs or the rest of the system, but the underlying system API is ABI stable.

Reply:
(This does have some interesting implications for Swift C++ interop on Windows, in that it means that user code would need to be using the same C/C++ runtime as the Swift runtime. Similarly if anything uses Swift runtime functions that return `malloc`-ed memory and so on.)